forch/faucetizer.py

```diff
@@ -1,49 +1,164 @@
 """Collect Faucet information and generate ACLs"""
 
+import argparse
 import logging
 import os
 import sys
+import yaml
 
 from forch.forchestrator import configure_logging
-from forch.utils import proto_dict
 from forch.utils import yaml_proto
 
 from forch.proto.network_state_pb2 import NetworkState
 
-LOGGER = logging.getLogger('topology')
+LOGGER = logging.getLogger('faucetizer')
 
 
 class Faucetizer:
     """Collect Faucet information and generate ACLs"""
-    def __init__(self):
-        self._network_state = {}
+    def __init__(self, output_file):
+        self._devices = {}
+        self._base_faucet_config = None
+        self._faucet_config = None
+        self._output_file = output_file
 
     def process_network_state(self, network_state):
-        """Process network state inputs"""
-        self._network_state = proto_dict(network_state, True)
+        """Process network state input"""
+        macs = set()
+        for mac, learning in network_state.device_mac_learnings.items():
+            macs.add(mac)
+            self._process_device_learning(mac, learning)
+        for mac, behavior in network_state.device_mac_behaviors.items():
+            macs.add(mac)
+            self._process_device_behavior(mac, behavior)
 
-        sorted_rules = list(self._network_state.get("named_match_rules", {}).keys())
-        sorted_rules.sort()
-        sys.stdout.write(f'{sorted_rules}\n')
+        self._faucetize(macs)
 
-        sorted_macs = list(self._network_state.get('device_mac_behaviors', {}).keys())
-        sorted_macs.sort()
-        sys.stdout.write(f'{sorted_macs}\n')
+    def _process_device_learning(self, mac, learning):
+        device_map = self._devices.setdefault(mac, {})
+        device_map['new_switch'] = learning.switch
+        device_map['new_port'] = learning.port
 
+    def _process_device_behavior(self, mac, behavior):
+        device_map = self._devices.setdefault(mac, {})
+        device_map['vlan'] = behavior.vid
+        device_map['role'] = behavior.role
 
-def load_network_state(base_dir_name):
+    def process_faucet_config(self, faucet_config):
+        """Process faucet config when base faucet config changes"""
+        self._base_faucet_config = faucet_config
+        self._faucet_config = faucet_config
+        self._faucetize()
+
+    def _faucetize(self, macs=None):
+        if not self._faucet_config:
+            return
+
+        if not macs:
+            macs = self._devices.keys()
+
+        commit = False
+
+        for mac in macs:
+            old_switch = self._devices.get(mac, {}).get('old_switch')
```

Comment:
replace with attributes not string keys

if necessary you can create a protobuf message for whatever this object is

Reply:
Added Device class for encapsulation